### ISSUE TITLE
failed Building production JavaScript and CSS bundles

### ISSUE THREAD

**Issue by walidcherhane** [gatsbyjs/gatsby#36304]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

The gatsby build failed for the wired error

ERROR #98123  WEBPACK

Generating JavaScript bundles failed

**Unknown option: .resourceQuery.**
**Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.**

File: src\pages\**

I see that its caused by _.resourceQuery_ but what the heck is that?? idk why
  

### Reproduction Link

https://github.com/pieh/i36304-minimal

### Steps to Reproduce

`npm run build`

### Expected Result

A successful build :/

### Actual Result

 ERROR #98123  WEBPACK

Generating JavaScript bundles failed

Unknown option: .resourceQuery. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.

File: src\pages\contact\contact1.tsx

 ERROR #98123  WEBPACK

Generating JavaScript bundles failed

Unknown option: .resourceQuery. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.

File: src\pages\contact\contact2.tsx

 ERROR #98123  WEBPACK

Generating JavaScript bundles failed

Unknown option: .resourceQuery. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.

File: src\pages\index.tsx

 ERROR #98123  WEBPACK

Generating JavaScript bundles failed

Unknown option: .resourceQuery. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.

File: src\pages\tours\index.tsx

 ERROR #98123  WEBPACK

Generating JavaScript bundles failed

Unknown option: .resourceQuery. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.

File: src\pages\404.tsx

 ERROR #98123  WEBPACK

Generating JavaScript bundles failed

Unknown option: .resourceQuery. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.

File: src\templates\tour-details.tsx

 ERROR #98123  WEBPACK

Generating JavaScript bundles failed

Unknown option: .resourceQuery. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.

File: src\templates\tours_by_categorie.tsx

 ERROR #98123  WEBPACK

Generating JavaScript bundles failed

Unknown option: .resourceQuery. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.

File: src\templates\tours_by_city.tsx

not finished Running gatsby-plugin-sharp.IMAGE_PROCESSING jobs - 39.494s

### Environment

```shell
System:
    OS: Windows 10 10.0.19043
    CPU: (4) x64 Intel(R) Core(TM) i5-6300U CPU @ 2.40GHz
  Binaries:
    Node: 16.14.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.19 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 8.3.1 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Spartan (44.19041.1266.0), Chromium (103.0.1264.77)
  npmPackages:
    gatsby: 4.17.2 => 4.17.2 
    gatsby-image: 3.11.0 => 3.11.0 
    gatsby-plugin-image: 2.18.0 => 2.18.0 
    gatsby-plugin-postcss: 5.18.0 => 5.18.0 
    gatsby-plugin-sharp: 4.18.0 => 4.18.0 
    gatsby-source-contentful: 7.16.0 => 7.16.0 
    gatsby-source-filesystem: 4.18.0 => 4.18.0 
    gatsby-transformer-json: 4.18.0 => 4.18.0 
    gatsby-transformer-remark: 5.18.0 => 5.18.0 
    gatsby-transformer-sharp: 4.18.0 => 4.18.0 
  npmGlobalPackages:
    gatsby-cli: 4.17.1
```

**Comment by pieh** [comment]
The reproduction link leads to "Page not found" - is this repo private?

**Comment by walidcherhane** [comment]
> The reproduction link leads to "Page not found" - is this repo private?

Yes forget to make it public : please check now :)

**Comment by LekoArts** [comment]
Please provide a [minimal reproduction][reproduction] and not your full project. It contains too much unrelated information and packages. You're also not using the latest version of Gatsby.

[reproduction]: https://gatsby.dev/reproduction

**Comment by Nat-friedman** [comment]
i get the same issue using the following versions 
    "gatsby": "^4.4.0",
    "gatsby-cli": "^4.20.0",
    "gatsby-plugin-create-client-paths": "^4.4.0",
    "gatsby-plugin-manifest": "^4.4.0",
    "gatsby-plugin-react-helmet": "^5.4.0",
    "gatsby-plugin-sharp": "^4.4.0",
    "gatsby-plugin-svgr": "^3.0.0-beta.0",
    "gatsby-plugin-typescript": "^4.5.1",
    "gatsby-transformer-sharp": "^4.4.0",

**Comment by pieh** [comment]
@Nat-friedman are those failures on `.ts(x)` files only as well?

**Comment by pieh** [comment]
https://github.com/pieh/i36304-minimal is the minimal repro

trigger that I found is using new(ish) `gatsby-plugin-typescript` and "older" `gatsby`.

In particular changes to `gatsby-plugin-typescript` introduced in https://github.com/gatsbyjs/gatsby/pull/35980/files#diff-2a73f504aeed7850c45f2030e878b4d8ea5f3b412233d42a44047d6f2690a8bd are not backward compatible, as the options we pass there on older gatsby versions are being router to babel itself because older gatsby didn't destructure `resourceQuery` in https://github.com/gatsbyjs/gatsby/blob/9df3cc2e381a207b927f96741f73f67cc35a58ec/packages/gatsby/src/utils/babel-loader.js#L34-L42 and therefore it was passing it as option to `loader` and not as `custom` option in https://github.com/gatsbyjs/gatsby/blob/9df3cc2e381a207b927f96741f73f67cc35a58ec/packages/gatsby/src/utils/babel-loader.js#L49-L68

This is a bit tricky, to make this backward compatible `gatsby-plugin-typescript` would need to know if those options are supported by gatsby or not

**Comment by Nat-friedman** [comment]
@pieh it's a mix of .js and tsx files :
failed Building production JavaScript and CSS bundles - 29.837s

 ERROR #98123  WEBPACK

Generating JavaScript bundles failed

Unknown option: .resourceQuery. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.

File: src\pages\invite.tsx


 ERROR #98123  WEBPACK

Generating JavaScript bundles failed

Unknown option: .resourceQuery. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.

File: src\pages\login.tsx

                                                                                                                                                                                                                            
 ERROR #98123  WEBPACK

Generating JavaScript bundles failed

Unknown option: .resourceQuery. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.

File: src\pages\reset-password.tsx


 ERROR #98123  WEBPACK

Generating JavaScript bundles failed

Unknown option: .resourceQuery. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.

File: src\pages\a.js


 ERROR #98123  WEBPACK

Generating JavaScript bundles failed

Unknown option: .resourceQuery. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.

File: src\pages\index.js


 ERROR #98123  WEBPACK

Generating JavaScript bundles failed

Unknown option: .resourceQuery. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.

File: src\pages\404.js

I'll take a look at the above comment and see if that fixes our issues, cheers 👍

**Comment by pieh** [comment]
Ok, I got a hang on why `.js` files are also showing up there. On top of my previous comment - with the mix of both `.js` and `.tsx` files it might happen semi-randomly depending on order in which webpack will compile files.

We have babel config cache (otherwise we would be validating that config for each file separately which is expensive), and depending on order it will use either problem config or proper one. In my minimal repro I checked and when I checked `404.tsx` and `index.js` - it was fine, but if I had `404.js` and `index.tsx` I was getting failures on both files.

Good news is that this problem will also be resolved with the fix to `gatsby-plugin-typescript` to make it backward compatible ( tho we have to figure out how to best do it retroactively ).

---

In the meantime - viable workaround is to:
 - either upgrade `gatsby`
 - pin `gatsby-plugin-typescript` to version corresponding to `gatsby` one (more or less) - if you don't install this plugin manually (gatsby ships with it by default) you will have to use something like yarn resolutions or `npm-force-resolutions` depending on your package manager of choice)
